server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -0,0 +1,1702 @@
+                    final org.elasticsearch.indices.recovery.RecoveryState recoveryState = internalCluster().getInstance(IndicesService.class, failingNode)
+                        .getShardOrNull(new ShardId(resolveIndex(indexName), 0)).recoveryState();
+                    assertThat(recoveryState.getTranslog().recoveredOperations(), equalTo(recoveryState.getTranslog().totalLocal()));
+                    if (startRecoveryRequestFuture.isDone()) {
+                        assertThat(recoveryState.getTranslog().totalLocal(), equalTo(0));
+                        recoveryState.getTranslog().totalLocal(localRecoveredOps.get());
+                        recoveryState.getTranslog().incrementRecoveredOperations(localRecoveredOps.get());
+                    } else {
+                        localRecoveredOps.set(recoveryState.getTranslog().totalLocal());
+                        startRecoveryRequestFuture.onResponse((StartRecoveryRequest) request);
+                    }
+                }
+                if (action.equals(PeerRecoveryTargetService.Actions.FILE_CHUNK)) {
+                    RetentionLeases retentionLeases = internalCluster().getInstance(IndicesService.class, node)
+                        .indexServiceSafe(resolveIndex(indexName))
+                        .getShard(0).getRetentionLeases();
+                    throw new AssertionError("expect an operation-based recovery:" +
+                                             "retention leases" + Strings.toString(retentionLeases) + "]");
+                }
+                connection.sendRequest(requestId, action, request, options);
+            });
+        }
+        IndexShard shard = internalCluster().getInstance(IndicesService.class, failingNode)
+            .getShardOrNull(new ShardId(resolveIndex(indexName), 0));
+        final long lastSyncedGlobalCheckpoint = shard.getLastSyncedGlobalCheckpoint();
+        final long localCheckpointOfSafeCommit;
+        try(Engine.IndexCommitRef safeCommitRef = shard.acquireSafeIndexCommit()){
+            localCheckpointOfSafeCommit =
+                SequenceNumbers.loadSeqNoInfoFromLuceneCommit(safeCommitRef.getIndexCommit().getUserData().entrySet()).localCheckpoint;
+        }
+        final long maxSeqNo = shard.seqNoStats().getMaxSeqNo();
+        shard.failShard("test", new IOException("simulated"));
+        StartRecoveryRequest startRecoveryRequest = startRecoveryRequestFuture.actionGet();
+        logger.info("--> start recovery request: starting seq_no {}, commit {}", startRecoveryRequest.startingSeqNo(),
+                    startRecoveryRequest.metadataSnapshot().getCommitUserData());
+        SequenceNumbers.CommitInfo commitInfoAfterLocalRecovery = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(
+            startRecoveryRequest.metadataSnapshot().getCommitUserData().entrySet());
+        assertThat(commitInfoAfterLocalRecovery.localCheckpoint, equalTo(lastSyncedGlobalCheckpoint));
```

Comment:
This assertion would fail without recent fix https://github.com/crate/crate/commit/abd07cad1bfea9f59f6c80955507582be9238bae